src/Serializer/DefaultHydrator.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Serializer;
+
+use Patchlevel\EventSourcing\Attribute\Normalize;
+use ReflectionClass;
+use ReflectionProperty;
+use RuntimeException;
+use TypeError;
+
+use function array_key_exists;
+use function assert;
+use function is_string;
+
+final class DefaultHydrator implements Hydrator
+{
+    /** @var array<class-string, array<string, array{reflection: ReflectionProperty, fieldName: string, normalizer: ?Normalizer}>> */
+    private static array $propertyMetadataCache = [];
+
+    /** @var array<class-string, ReflectionClass> */
+    private static array $reflectionClassCache = [];
+
+    /**
+     * @param class-string<T>      $class
+     * @param array<string, mixed> $data
+     *
+     * @return T
+     *
+     * @template T of object
+     */
+    public function hydrate(string $class, array $data): object
+    {
+        $object = self::reflectionClass($class)->newInstanceWithoutConstructor();
+        assert($object instanceof $class);
+
+        $metadata = self::propertyMetadata($class);
+
+        foreach ($metadata as $propertyMetadata) {
+            $value = $data[$propertyMetadata['fieldName']] ?? null;
+
+            if ($propertyMetadata['normalizer']) {
+                $value = $propertyMetadata['normalizer']->denormalize($value);
+            }
+
+            try {
+                $propertyMetadata['reflection']->setValue($object, $value);
+            } catch (TypeError $error) {
+                throw new TypeMismatch($error->getMessage(), 0, $error);
+            }
+        }
+
+        return $object;
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    public function extract(object $object): array
+    {
+        $metadata = self::propertyMetadata($object::class);
+
+        $data = [];
+
+        foreach ($metadata as $propertyMetadata) {
+            $value = $propertyMetadata['reflection']->getValue($object);
+
+            if ($propertyMetadata['normalizer']) {
+                $value = $propertyMetadata['normalizer']->normalize($value);
+            }
+
+            $data[$propertyMetadata['fieldName']] = $value;
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param class-string $class
+     *
+     * @return array<string, array{reflection: ReflectionProperty, fieldName: string, normalizer: ?Normalizer}>
+     */
+    private static function propertyMetadata(string $class): array
+    {
+        if (array_key_exists($class, self::$propertyMetadataCache)) {
+            return self::$propertyMetadataCache[$class];
+        }
+
+        $reflectionClass = self::reflectionClass($class);
+        $metadata = [];
+
+        foreach ($reflectionClass->getProperties() as $property) {
+            $property->setAccessible(true);
+
+            $normalizer = null;
+            $attributeReflectionList = $property->getAttributes(Normalize::class);
+
+            if ($attributeReflectionList !== []) {
```

Comment:
count should be faster i think

Reply:
I really don't know if it's faster or slower. From a feeling I would say that my version should be faster because it doesn't "count" here. Even if opcache optimization happens here, the optimization should also be applicable to my variant. I have searched for it but haven't really found anything. Are you 100% sure it is faster?